New FPGA reference design: merge sort

tyoungsc opened the pull request:
Signed-off-by: tyoungsc <tanner.young-schultz@intel.com>

## Description
Creating new FPGA reference design; merge sort (NOTE FROM REVIEWER: please add more detail here for the PR to the main repo)

## External Dependencies
None

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
- [x] Command Line
No method to test this with our infra yet, so I tested all 3 flows manually.

tyoungsc committed:
First commit for merge sort design

Signed-off-by: tyoungsc <tanner.young-schultz@intel.com>

tyoungsc committed:
Cleaned up MergeSort.hpp a bit using defines

Signed-off-by: tyoungsc <tanner.young-schultz@intel.com>

tyoungsc committed:
Formatting/comments

tyoungsc committed:
Reduced area by connecting Partition unit (previously called Shuffle) to
only the first merge unit
Changed Shuffle to Partition
Updated pictures and README
Used existing pipe_array code (instead of using my own)
Code cleanup and comments

Tested in emulation and reports, doing a HW build now

tyoungsc committed:
Fixed bug with no-USM BSPs in Produce kernel
Slight code cleanup

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md:
```diff
@@ -0,0 +1,205 @@
+<br>
+
+**Performance**
+The performance data below was gathered using the Intel® FPGA PAC D5005 (with Intel Stratix® 10 SX) sorting `2^24=16777216` elements using 1-16 merge units and the best throughput across 5 seeds.
```
Do you think we should include which version of the compiler was used to achieve these results?  My thinking here is to avoid having to update these numbers in the future.

tyoungsc replied:
I think since the designs are tagged for each release, it would be weird to not redo this for every release (I have a script to do this with regtests)

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md:
```diff
@@ -0,0 +1,205 @@
+|`Partition.hpp`                 | The `Partition` kernel
+|`pipe_array.hpp`                | Header file containing the definition of an array of pipes
+|`pipe_array_internal.hpp`       | Helper for pipe_array.hpp
+|`Produce.hpp`                   | The `Produce` kernel for the merge unit
```
Your descriptions for Consume, Merge, Produce, and Partition are . . . unhelpful :-).  Is it worth adding one sentence of useful description instead of just repeating the name of the kernel?

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md:
```diff
@@ -0,0 +1,205 @@
+### Merge Sort Details
+This section will describe how the merge sort design is layed out and how it takes advantage of the spatial computing of the FPGA. <br/>
+
+The figure below shows the conceptual view of the merge sort design to the user. The user streams data into a SYCL pipe and, after some delay, the element are streamed out of the sorter, in sorted order. The number of elements the merge sort design sorts can be a runtime parameter, but it must be a power of 2. However, this restriction can be worked around by padding the input stream with min/max ...
```
"after some delay, the elementS are . . ." plural

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md:
```diff
@@ -0,0 +1,205 @@
+
+<img src="sort_api.png" alt="sort_api" width="600"/>
+
+This basis of the merge sort design is what we call a *merge unit*, which is shown in the figure below. A single merge unit streams in sorted lists of size `count` in parallel and merges them into a sorted list of size `2*count`. The lists are streamed in by the `Produce` kernel, which can be configured to stream the data from a SYCL pipe or device memory (e.g., DDR or HBM). This configuration ca ...
```
"A single merge unit streams in TWO sorted lists..."

"...and merges them into a SINGLE sorted list of..."

"...or at runtime with a dynamic parameter that changes over time." - I don't understand what this means?  I think just saying runtime parameter implies it is 'dynamic', and that it can change (over time?).  So I think this is triple redundant, unless you're trying to convey some additional information?

"...can stream data TO either a ..."

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md:
```diff
@@ -0,0 +1,205 @@
+
+<img src="basic_runtime_graph.png" alt="basic_runtime_graph" width="800"/>
+
+To improve performance, the merge sort design accepts a template parameter `units` which allows one to instantiate multiple instances of the merge unit, as shown in the figure below. Choosing the number of units is an area-performance tradeoff (note: the number of instantiated merge units must be a power of 2). In this design, each merge unit sorts a partition of the input data of size `N/units`. ...
```
"feeds the data to producers of merge unit 0 (alternating between the)" - alternating between the what?

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md:
```diff
@@ -0,0 +1,205 @@
+### Performance disclaimers
+Tests document performance of components on a particular test, in specific systems. Differences in hardware, software, or configuration will affect actual performance. Consult other sources of information to evaluate performance as you consider your purchase.  For more complete information about performance and benchmark results, visit [www.intel.com/benchmarks](www.intel.com/benchmarks).
+
+Performance results are based on testing as of July 29, 2020 and may not reflect all publicly available security updates.  See configuration disclosure for details.  No product or component can be absolutely secure.
```
Should this date be changed?

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md:
```diff
@@ -0,0 +1,205 @@
+
+Intel technologies’ features and benefits depend on system configuration and may require enabled hardware, software or service activation. Performance varies depending on system configuration. Check with your system manufacturer or retailer or learn more at [intel.com](www.intel.com).
+
+The performance was measured by Intel on July 29, 2020.
```
Do we need the date twice?  I'm assuming this is just copied from some standard text, but it bothers me.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md:
```diff
@@ -0,0 +1,205 @@
+
+Intel and the Intel logo are trademarks of Intel Corporation or its subsidiaries in the U.S. and/or other countries.
+
+(C) Intel Corporation.
```
Can't comment on a binary file, so commenting here.

You've included a file called 'basic.png' which I don't think is actually used (and if it is - it still has the 'shuffle' name, so please fix that).

In 'parallel_tree.png', it looks like the pipe output from the first kernel goes through the partition block, please draw the arrow around that block, or just make the partition block smaller or shift it left.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/third-party-programs.txt:
```diff
@@ -0,0 +1,253 @@
+oneAPI Code Samples - Third Party Programs File
```
This file does not need to be included in designs that are not mentioned in the file, so please remove it.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/sample.json:
```diff
@@ -0,0 +1,61 @@
+{
+  "guid": "AF215180-ECB8-4414-8049-D89E4882576D",
```
Please confirm you created a new guid for this design and didn't just cut-and-paste this from another design.

tyoungsc replied:
Confirmed

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/src/CMakeLists.txt:
```diff
@@ -0,0 +1,108 @@
+
+# This is a Windows-specific flag that enables error handling in host code
+if(WIN32)
+    set(WIN_FLAG "-EHsc")
```
/EHsc

tyoungsc replied:
I thought Alain was changing all these...?

mdbtucker replied:
Yeah, that was a mistake (by me), we don't actually want to change these, they should still be /

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/src/CMakeLists.txt:
```diff
@@ -0,0 +1,108 @@
+separate_arguments(CMAKE_CXX_FLAGS_LIST)
+set(HARDWARE_LINK_FLAGS_LIST ${HARDWARE_LINK_FLAGS})
+separate_arguments(HARDWARE_LINK_FLAGS_LIST)
+add_custom_command(OUTPUT ${FPGA_EARLY_IMAGE}
```
Please update your CMake to match the newest standard template.  I think the Generate Report section is the only change from what you have, but please double check.

tyoungsc replied:
I copied the one from the CRR design in the tip of GitHub. I will base it off MVDR.

mdbtucker replied:
Base it off of this: https://github.com/oneapi-src/oneAPI-samples/pull/492 (except for changing /EHsc to -EHsc - don't do that, as mentioned elsewhere)

mdbtucker reviewed (commented):
Done reviewing all non-source (.hpp .cpp) files

tyoungsc committed:
README update after Mike's review
CMake update
Removed line from samples.json that was not necessary
Deleted unused files

tyoungsc committed:
Changed Windows specific flag to use '/' instead of '-'

tyoungsc committed:
CMake update

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md:
```diff
@@ -0,0 +1,198 @@
+# Merge Sort
```
Please have a look at the changes Audrey made here: https://github.com/oneapi-src/oneAPI-samples/pull/507/ and replicate those changes in your README

mdbtucker replied:
I assume you did this?  If so, please 'resolve' this conversation.

tyoungsc committed:
Removed all pointer arithmetic for offseting to the inside of kernels
(to avoid runtime issue)

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/src/Misc.hpp:
```diff
@@ -0,0 +1,39 @@
+#ifndef __MISC_HPP__
```
Misc is a bad name for this file.  How about 'static_math'?

tyoungsc replied:
Ran clang-format, renamed Misc as you suggested.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/src/Misc.hpp:
```diff
@@ -0,0 +1,39 @@
+constexpr T Log2(T n) {
+  static_assert(std::is_integral<T>::value);
+  static_assert(std::is_unsigned<T>::value);
+  return ((n < 2) ? T(0) : T(1) + Log2(n / 2));
```
Recursive functions are not allowed in kernels.  Doesn't really matter since you're not using this in a kernel, but why not re-write this as a loop so it could be used inside kernels?  See https://github.com/oneapi-src/oneAPI-samples/blob/master/DirectProgramming/DPC%2B%2BFPGA/ReferenceDesigns/mvdr_beamforming/src/StreamingQRD.hpp

tyoungsc replied:
Done.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/src/merge_sort.cpp:
```diff
@@ -0,0 +1,334 @@
+
+// the number of merge units, must be a power of 2
+#ifndef MERGE_UNITS
+#ifdef FPGA_EMULATOR
```
Is it common practice to modify behaviour inside the source code based on whether we are running emulation or not?  I'm not a big fan of this practice, I'd rather see changes like that happen in Makefiles / cmake files.  The danger is you don't realize you're not emulating the code you think you are.

What happens if we emulate 16 merge units? If the run takes too long, wouldn't it make sense to just sort fewer things in emulation, rather than changing the structure of the code?

I'm open to being convinced that the way you've done it is ok.

tyoungsc replied:
Fixed. This was a biproduct of my debugging.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/src/merge_sort.cpp:
```diff
@@ -0,0 +1,334 @@
+  std::vector<ValueT> in_vec(count), out_vec(count), ref(count);
+
+  // generate some random input data
+  srand(time(NULL));
```
I suggest you make the seed a command line argument, so results are reproducible.  If you want to keep randomization based on time an option, you could do that if no command line seed is provided.

tyoungsc replied:
Done.

mdbtucker reviewed (commented):
Please confirm you've run clang-format on all code.

tyoungsc committed:
Big change to example design: encorperate a new multi-element per cycle
merge unit, use a bitonic sorter on the input, rather than a simple
partition.

Updated README and pictures to fit the new design.

Addressed Mike's most recent review comments

tyoungsc committed:
Code cleanup: changed some variable names and comments

tyoungsc committed:
README update

tyoungsc committed:
Picture update

tyoungsc committed:
Comments

tyoungsc committed:
Code cleanup and comments

jessicadavies-intel commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md:
```diff
@@ -0,0 +1,206 @@
+
+<img src="merge_unit.png" alt="merge_unit" width="600"/>
+
+A single merge unit requires `lg(N)` iterations to sort `N` elements. This requires the host enqueuing `lg(N)` iterations of the merge unit kernels that merge sublists of size {`1`, `2`, `4`, ...} into larger lists of size {`2`, `4`, `8`, ...}, respectively. This results in a timeline that looks like the figure below.
```
"This requires the host enqueuing `lg(N)` iterations of the merge unit kernels"
Should it be "merge unit kernel" instead of "kernels" here?

tyoungsc replied:
Each merge unit has 4 kernels.

jessicadavies-intel commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md:
```diff
@@ -0,0 +1,206 @@
+
+<img src="basic_runtime_graph.png" alt="basic_runtime_graph" width="800"/>
+
+To achieve SIMD level parallelism, we enhance the merge unit to merge `k` elements per cycle. The figure below shows how this is done. We will assume that we are sorting from largest to smallest, but the logic is very similar for sorting largest to smallest and is easily configurable at compile time in this design. <br/>
```
"but the logic is very similar for sorting largest to smallest" should be "smallest to largest".

tyoungsc replied:
Fixed. It should be "assuming smallest to largest.... similar for soring largest to smallest".

jessicadavies-intel commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md:
```diff
@@ -0,0 +1,206 @@
+
+To achieve thread level parallelism, the merge sort design accepts a template parameter, `units`, which allows one to instantiate multiple instances of the merge unit, as shown in the figure below. Before the merge units start, the incoming data coming from the input pipe is sent through a bitonic sorting network and written to the temporary buffer partitions in device memory. This sorting networ ...
+
+<img src="parallel_tree_bitonic_k-way.png" alt="parallel_tree_bitonic_k-way" width="800"/>
```
Why do the merge units each have an arrow showing that they write to DDR? I don't see where this is mentioned in the Readme, did I miss it?

tyoungsc replied:
Because the write to temp read buffers in device memory.
"The lists are streamed in from device memory (e.g., DDR or HBM) by two `Produce` kernels. The `Consumer` kernel can stream data to either a SYCL pipe or to device memory."

jessicadavies-intel commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md:
```diff
@@ -0,0 +1,206 @@
+
+<img src="parallel_tree_bitonic_k-way.png" alt="parallel_tree_bitonic_k-way" width="800"/>
+
+After the merge units sort their `N/units`-sized partition, the partitions must be reduced into a single sorted list. There are two options to do this: (1) reuse the merge units to perform `lg(units)` more iterations to sort the partitions, or (2) create a merge tree to reduce the partitions into a single sorted list. Option (1) saves area at the expense of performance, since it has to perform ad ...
```
"that feed the merge tree." should be "that feeds the merge tree."

jessicadavies-intel commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md:
```diff
@@ -0,0 +1,206 @@
+
+<img src="k-way_merge_unit.png" alt="way_merge_unit" width="900"/>
+
+To achieve thread level parallelism, the merge sort design accepts a template parameter, `units`, which allows one to instantiate multiple instances of the merge unit, as shown in the figure below. Before the merge units start, the incoming data coming from the input pipe is sent through a bitonic sorting network and written to the temporary buffer partitions in device memory. This sorting networ ...
```
"Choosing the number of merge units is an area-performance tradeoff"
The table of # units vs. throughput appeared at the top of the Readme, but it would be nice to repeat and/or discuss it here. Also, can area be added to this table?

tyoungsc replied:
I am debating this. I don't feel like upkeeping a table full or performance/area numbers.

tyoungsc committed:
Small change to README

tyoungsc committed:
Updating design output and README

tyoungsc committed:
Simplified Produce kernel

tyoungsc committed:
Allowed SORT_WIDTH to be 1 (1 element per cycle)
Improved Merge kernel for case where SORT_WIDTH=1 with shannonization

tyoungsc committed:
Code cleanup
        Variable renaming
        Comments
        Grammar

tyoungsc committed:
Comments, formatting, README grammar changes, and general cleanup :)

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/src/static_math.hpp:
```diff
@@ -1,5 +1,12 @@
+#ifndef __STATICMATH_HPP__
+#define __STATICMATH_HPP__
+
+#define SWAP(a, b)  \
```
Does this function really belong in 'static math', since it appears to be neither math nor static?

tyoungsc replied:
Can I rename the file "common_utilities.hpp"?? Would that address the problem?

tyoungsc replied:
File renamed to impu_math.hpp

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/src/static_math.hpp:
```diff
@@ -1,5 +1,12 @@
-#define __MISC_HPP__
+#ifndef __STATICMATH_HPP__
+#define __STATICMATH_HPP__
+
```
Should we put these functions in a namespace?  I'm thinking ahead in case we want to elevate this file to a common header library.

tyoungsc replied:
Probably. What should the name of the namespace be? What if I rename the file common_utilities.hpp and make the namespace "common_util"??

Maybe sycl::fpga::common_util ?? Or something?

tyoungsc replied:
Used namespace impu::math

mdbtucker reviewed (commented):
Not many comments, I don't know if the code is in really good shape, or I'm feeling lenient, or I'm just lazy :-).

tyoungsc committed:
Renamed all files to match google style
Renamed main file to main.cpp
Used impu namespace in unrolledloop, pipearray, and static_math
Renamed static_math.hpp to impu_math.hpp
Updated README, Windows VS files, and CMake with these changes

tyoungsc committed:
README updates
Comments

tyoungsc committed:
Changed filenames to use underscores between words

tyoungsc committed:
Added 'pipe' namespace to impu namespace for pipe utilities

mdbtucker reviewed (approved):
Looks good, you can go ahead and create a PR to the main repo.
